state/postgresql/v2/postgresql.go

```diff
@@ -0,0 +1,506 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package postgresql
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/jackc/pgx/v5/pgxpool"
+
+	pginterfaces "github.com/dapr/components-contrib/common/component/postgresql/interfaces"
+	pgtransactions "github.com/dapr/components-contrib/common/component/postgresql/transactions"
+	sqlinternal "github.com/dapr/components-contrib/common/component/sql"
+	pgmigrations "github.com/dapr/components-contrib/common/component/sql/migrations/postgres"
+	"github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/components-contrib/state"
+	stateutils "github.com/dapr/components-contrib/state/utils"
+	"github.com/dapr/kit/logger"
+)
+
+// PostgreSQL state store.
+type PostgreSQL struct {
+	state.BulkStore
+
+	logger   logger.Logger
+	metadata pgMetadata
+	db       pginterfaces.PGXPoolConn
+
+	gc sqlinternal.GarbageCollector
+
+	enableAzureAD bool
+}
+
+type Options struct {
+	// Disables support for authenticating with Azure AD
+	// This should be set to "false" when targeting different databases than PostgreSQL (such as CockroachDB)
+	NoAzureAD bool
+}
+
+// NewPostgreSQLStateStore creates a new instance of PostgreSQL state store with the default options.
+func NewPostgreSQLStateStore(logger logger.Logger) state.Store {
+	return NewPostgreSQLStateStoreWithOptions(logger, Options{})
+}
+
+// NewPostgreSQLStateStoreWithOptions creates a new instance of PostgreSQL state store with options.
+func NewPostgreSQLStateStoreWithOptions(logger logger.Logger, opts Options) state.Store {
+	s := &PostgreSQL{
+		logger:        logger,
+		enableAzureAD: !opts.NoAzureAD,
+	}
+	s.BulkStore = state.NewDefaultBulkStore(s)
+	return s
+}
+
+// Init sets up Postgres connection and performs migrations
+func (p *PostgreSQL) Init(ctx context.Context, meta state.Metadata) error {
```

Comment:
Can we include somewhere here (apart from the docs) - the diff with v1 and what are the benefits of using v2 - just, so that a developer can upfront see the reason for a v2 here.

Reply:
Done, in the comment on the NewPostgreSQLStateStore method